### ISSUE TITLE
Infinite loop in the C function medcouple()

### ISSUE THREAD

**Issue by pach318** [FilippoBovo/robustats#6]
While using the module to calculate medcouple, I encountered a data sequence that would cause the process to consume 100% CPU and not return. This is traced to the C function medcouple(), which goes into an infinite loop when processing the specific data sequence.

After some debugging, it looks like the root cause is that the where_h_greater_than_u() and where_h_less_than_u() functions are not passing the correct epsilon to the h_kernel() function. (The h_kernel() function should be using the "epsilon2" to compare the equality between "a" and "b".)

I will submit a pull request with the fix and a test case containing the specific data sequence.

**Comment by pach318** [comment]
The pull request #7  includes two commits. The first commit changes the where_h_greater_than_u() and where_h_less_than_u() functions so that they pass the correct epsilon ("epsilon2") to h_kernel(), while still using the original epsilon ("wm_epsilon") for their own comparisons. The second commit adds a unittest case for the specific data sequence (note: running the new test case without the fix would result in an infinite loop, and the process will need to be killed manually).

Please check if the changes can be applied. Thanks.

**Comment by FilippoBovo** [comment]
@pach318, thank you for the code fix.

Before I merge to master, could you please let me know your reasoning for using `epsilon2` in functions `where_h_greater_than_u` and `where_h_less_than_u` instead of `wm_epsilon`?

In addition, when you debugged the C code with the array that you used in the new unit test, do you remember which element(s) of the array caused the code to loop?

**Comment by pach318** [comment]
When tracing the algorithm, I noticed that among the four places that use the `h_kernel()` function, the above two functions are passing `wm_epsilon` (which can change in each iteration of the main loop) to `h_kernel()`, while the other two places (in `medcouple()` itself) are passing `epsilon2`.

I checked some other available implementations:
- The [C++ implementation](http://inversethought.com/hg/medcouple/file/default/jmedcouple.c%2B%2B) listed on the [medcouple Wikipedia page](https://en.wikipedia.org/wiki/Medcouple)
- A [Python implementation](http://inversethought.com/hg/medcouple/file/tip/medcouple.py) by the same author as above
- The `robustbase` package in R

As far as I can see the `h_kernel()` (or equivalent) function in these implementations always use the `epsilon2` (or equivalent) for the comparison between `a` and `b`.

I tested the specific data sequence using the C++ and the Python implementations listed above, plus three "slow" implementations: `statsmodels.stats.stattools.medcouple()`, the Matlab implementation listed on Wikipedia, and an Octave implementation (by the same author as the C++ one above). The result from these five implementations are the same (at least to 16 digits), and applying the epsilon change here also produces the same result. So at least for this specific sequence the result should be correct.

Regarding the data sequence, I did try removing some elements, but removing any one (of the ones I tried) would make the problem disappear, which would seem to indicate that the issue is not caused by a single particular element. However, I did not extensively test various combinations of the array elements, and I would imagine one should certainly be able to find a shorter sequence that can trigger the issue.
